File: backend/app/api/routes/files.py
from fastapi import APIRouter, UploadFile, File, Depends, HTTPException, status
from fastapi.responses import FileResponse
import os
from typing import List
from starlette.requests import Request
from .auth import get_current_username

router = APIRouter(prefix="/files", tags=["Files"])
UPLOAD_DIR = "uploads"
os.makedirs(UPLOAD_DIR, exist_ok=True)

@router.post("/upload")
async def upload_file(file: UploadFile = File(...), username: str = Depends(get_current_username)):
    try:
        file_path = os.path.join(UPLOAD_DIR, file.filename)
        with open(file_path, "wb") as f:
            f.write(await file.read())
        return {"status": "success", "filename": file.filename}
    except Exception as e:
        raise HTTPException(500, str(e))

@router.get("/")
async def list_files(username: str = Depends(get_current_username)) -> List[str]:
    return os.listdir(UPLOAD_DIR)

@router.get("/{filename}")
async def view_file(filename: str, request: Request, username: str = Depends(get_current_username)):
    file_path = os.path.join(UPLOAD_DIR, filename)
    if not os.path.exists(file_path):
        raise HTTPException(404, "File not found")

    ext = filename.lower().split('.')[-1]
    if ext in ['png', 'jpg', 'jpeg', 'gif']:
        return FileResponse(file_path, media_type=f"image/{ext}")
    elif ext == 'pdf':
        return FileResponse(file_path, media_type="application/pdf")
    elif ext in ['txt', 'csv', 'html', 'htm', 'xml', 'json']:
        return FileResponse(file_path, media_type="text/plain")
    elif ext in ['mp4', 'webm']:
        return FileResponse(file_path, media_type=f"video/{ext}")
    elif ext in ['mp3', 'wav', 'ogg']:
        return FileResponse(file_path, media_type=f"audio/{ext}")
    else:
        return FileResponse(
            file_path,
            media_type="application/octet-stream",
            filename=filename,
            headers={"Content-Disposition": f"attachment; filename={filename}"}
        )

@router.get("/download/{filename}")
async def download_file(filename: str, username: str = Depends(get_current_username)):
    file_path = os.path.join(UPLOAD_DIR, filename)
    if not os.path.exists(file_path):
        raise HTTPException(404, "File not found")
    return FileResponse(file_path, media_type="application/octet-stream", filename=filename)

@router.delete("/{filename}")
async def delete_file(filename: str, username: str = Depends(get_current_username)):
    safe_filename = os.path.basename(filename)
    file_path = os.path.join(UPLOAD_DIR, safe_filename)
    if not os.path.exists(file_path):
        raise HTTPException(404, "File not found")
    os.remove(file_path)
    return {"status": "File deleted successfully"}

File: frontend/public/code/files.js
async function uploadFile() {
    const fileInput = document.getElementById('file-input');
    if (fileInput.files.length === 0) {
        alert("Please select a file to upload.");
        return;
    }

    const formData = new FormData();
    formData.append("file", fileInput.files[0]);

    try {
        const response = await fetch('/upload', {
            method: 'POST',
            body: formData,
            headers: {
                'Authorization': 'Basic ' + btoa('admin:admin')
            }
        });

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const data = await response.json();
        if (data.status === 'success') {
            alert("File uploaded successfully!");
            loadFiles();
            fileInput.value = '';
        }
    } catch (error) {
        console.error('Upload error:', error);
        alert(`Upload failed: ${error.message}`);
    }
}

async function loadFiles() {
    try {
        const response = await fetch('/files', {
            headers: {
                'Authorization': 'Basic ' + btoa('admin:admin')
            }
        });

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const files = await response.json();
        const fileList = document.getElementById('fileList');
        fileList.innerHTML = '';

        files.forEach(file => {
            const fileItem = createFileItem(file);
            fileList.appendChild(fileItem);
        });
    } catch (error) {
        console.error('Error loading files:', error);
        alert(`Failed to load files: ${error.message}`);
    }
}

function createFileItem(filename) {
    const li = document.createElement('li');
    li.className = 'file-item';
    li.innerHTML = `
        <span class="filename">${filename}</span>
        <div class="file-actions">
            <button class="view-btn">View</button>
            <button class="download-btn">Download</button>
            <button class="delete-btn">Delete</button>
        </div>
    `;

    li.querySelector('.view-btn').addEventListener('click', () => viewFile(filename));
    li.querySelector('.download-btn').addEventListener('click', () => downloadFile(filename));
    li.querySelector('.delete-btn').addEventListener('click', () => deleteFile(filename));

    return li;
}

function viewFile(filename) {
    const extension = filename.split('.').pop().toLowerCase();
    const viewableExtensions = ['png', 'jpg', 'jpeg', 'gif', 'pdf', 'txt', 'csv', 'html', 'htm', 'mp4', 'webm', 'mp3', 'wav', 'ogg'];

    if (viewableExtensions.includes(extension)) {
        const url = `/files/${encodeURIComponent(filename)}`;
        const auth = 'Basic ' + btoa('admin:admin');

        const form = document.createElement('form');
        form.method = 'GET';
        form.action = url;
        form.target = '_blank';

        const authInput = document.createElement('input');
        authInput.type = 'hidden';
        authInput.name = 'Authorization';
        authInput.value = auth;

        form.appendChild(authInput);
        document.body.appendChild(form);
        form.submit();
        document.body.removeChild(form);
    } else {
        alert('This file type cannot be viewed directly. Please download it.');
    }
}

async function downloadFile(filename) {
    try {
        const response = await fetch(`/files/${encodeURIComponent(filename)}`, {
            headers: {
                'Authorization': 'Basic ' + btoa('admin:admin')
            }
        });

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const blob = await response.blob();
        const url = window.URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = filename;
        document.body.appendChild(a);
        a.click();
        document.body.removeChild(a);
        window.URL.revokeObjectURL(url);
    } catch (error) {
        console.error('Download error:', error);
        alert(`Download failed: ${error.message}`);
    }
}

async function deleteFile(filename) {
    if (!confirm(`Are you sure you want to delete ${filename}?`)) return;

    try {
        const response = await fetch(`/files/${encodeURIComponent(filename)}`, {
            method: 'DELETE',
            headers: {
                'Authorization': 'Basic ' + btoa('admin:admin')
            }
        });

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const data = await response.json();
        if (data.status === 'File deleted successfully') {
            alert('File deleted successfully!');
            loadFiles();
        }
    } catch (error) {
        console.error('Delete error:', error);
        alert(`Delete failed: ${error.message}`);
    }
}
